problem_builder/tasks.py

```diff
@@ -113,4 +88,108 @@ def scan_for_blocks(block):
+    # Extract info for "Answer" and "Username" columns
+    # - Get all of the most recent student submissions for this block:
+    submissions = _get_submissions(course_key_str, block, user_id)
+
+    # - For each submission, look up student's username and answer:
+    for submission in submissions:
+        username = _get_username(submission, user_id)
+        answer = _get_answer(block, submission)
+
+        # Short-circuit if answer does not match search criteria
+        if not match_string.lower() in answer.lower():
+            continue
+
+        rows.append([section_name, subsection_name, unit_name, block_type, block_question, answer, username])
+
+    return rows
+
+
+def _get_context(block):
+    """
+    Return section, subsection, and unit names for `block`.
+    """
+    block_names_by_type = {}
+    block_iter = block
+    while block_iter:
+        block_iter_type = block_iter.scope_ids.block_type
+        block_names_by_type[block_iter_type] = block_iter.display_name_with_default
+        block_iter = block_iter.get_parent() if block_iter.parent else None
+    section_name = block_names_by_type.get('chapter', '')
+    subsection_name = block_names_by_type.get('sequential', '')
+    unit_name = block_names_by_type.get('vertical', '')
+    return section_name, subsection_name, unit_name
+
+
+def _get_type(block):
+    """
+    Return type of `block`.
+    """
+    return block.scope_ids.block_type
+
+
+def _get_submissions(course_key_str, block, user_id):
+    """
+    Return submissions for `block`.
+    """
+    # Load the actual student submissions for `block`.
+    # Note this requires one giant query that retrieves all student submissions for `block` at once.
+    block_id = unicode(block.scope_ids.usage_id.replace(branch=None, version_guid=None))
+    block_type = _get_type(block)
+    if not user_id:
+        return sub_api.get_all_submissions(course_key_str, block_id, block_type)
+    else:
+        student_dict = {
+            'student_id': user_id,
+            'item_id': block_id,
+            'course_id': course_key_str,
+            'item_type': block_type,
+        }
+        return sub_api.get_submissions(student_dict, limit=1)
+
+
+def _get_username(submission, user_id):
+    """
+    Return username of student who provided `submission`.
+    """
+    # If the student ID key doesn't exist, we're dealing with a single student and know the ID already.
+    student_id = submission.get('student_id', user_id)
+    return user_by_anonymous_id(student_id).username
+
+
+def _get_answer(block, submission):
+    """
+    Return answer associated with this `submission` to `block`.
+    """
+    answer = submission['answer']
+    try:
+        choices = block.children
+    except AttributeError:
+        pass
+    else:
+        for choice in choices:
+            choice_block = modulestore().get_item(choice)
+            if choice_block.value == answer:
+                answer = choice_block.content
+                break
```

Comment:
@itsjeyd Nice improvements! This is great. I only have one final comment, which is about fetching the choices above. I believe the choices will always be the same for each block, so we should avoid loading all the choices from the modulestore for each student.

You could:
1. Move the code that looks for choice children up one level to be in `_extract_data` outside of the per-student loop.
or 
2. Keep it here but cache the result (this is probably more efficient since it avoids looking up answers that won't be needed).

Actually, after looking into this I found out that you might be able to take advantage of `QuestionnaireAbstractBlock.get_submission_display` and to make this cleaner:
```python
def _get_answer(block, submission, answer_cache):  # Answer cache is a dict that is reset for each block.
    answer = submission['answer']
    if isinstance(block, QuestionnaireAbstractBlock):
        # Convert from answer ID to answer label
        if answer not in answer_cache:
            answer_cache[answer] = block.get_submission_display(answer)
        return answer_cache[answer]
    return answer
```